### ISSUE TITLE
Passing model geometry directly to a compute shader

### ISSUE THREAD

**Issue by rileyzzz** [Facepunch/sbox-issues#7116]
### What can't you do?

I'd love if we had a way to take an instance of a Model (i.e. something off asset.party) and grab the vertex and index buffers as ComputeBuffer objects (so that you don't have to compile the model with CPU access, and can pipe the already-on-the-GPU mesh data to be read by a compute shader).
Not sure how feasible this is from an engine standpoint, but it'd be super cool to download models off asset.party and generate SDFs/what have you right there on the GPU without requiring that people tick "Enable CPU Access".

### How would you like it to work?

Maybe some functions on Model, i.e.
`ComputeBuffer<Vertex> GetVertexBuffer(int mesh)`, `ComputeBuffer<int> GetIndexBuffer(int mesh)`
where `mesh` is in [0, Model.MeshCount)? Depending on how the vertex data is packaged up internally that might need to be tweaked to handle multiple streams etc., in which case maybe grabbing a compute buffer for a specific VertexAttributeType might be more appropriate?

If that's not possible within the engine, a different (weird, but maybe simpler?) option could be an overload for RenderAttributes.Set i.e.:
`RenderAttributes.Set(in string vertexbuffer, in string indexbuffer, Model model, int meshIndex = 0);`
which would do whatever necessary work to set the structured buffers internally, without requiring exposing the GPU VB/IB objects to managed code.

### What have you tried?

Ticking "Enable CPU Access" and uploading the buffers manually is the only real workaround afaik, but it stinks as it effectively limits you to the models you've compiled with that option.

**Comment by handsomematt** [comment]
This would be cool, it'd probably have to be a copy into a new buffer that can be accessed from a compute shader.

I don't see this being crazy hard, you'd probably need to know the vertex format yourself though.

Do any other engines have APIs for this?

**Comment by rileyzzz** [comment]
Seems like Unity has a pretty similar API to grab the GPU buffers:

https://docs.unity3d.com/2023.2/Documentation/ScriptReference/Mesh.GetVertexBuffer.html

Not an expert on it but it looks like they expose a few more functions to let you query the vertex layout (stride, attribute offsets, stream index)

https://docs.unity3d.com/2023.2/Documentation/ScriptReference/Mesh.GetVertexAttributeOffset.html

https://docs.unity3d.com/2023.2/Documentation/ScriptReference/Mesh.GetVertexAttributeStream.html

ComputeBuffer<> being generic could pose a problem as the generic type argument isn't really relevant. Seems like Unity just gives you an opaque GraphicsBuffer object, and they expect you'll put it in a ByteAddressBuffer and figure out the offsets manually in the compute shader.

**Comment by handsomematt** [comment]
Yeah I've just replaced ComputeBuffer<T> with GpuBuffer which can do those things.

The GetVertexBuffer() in Unity looks like direct access to the underlying buffer instead of a copy which is interesting. I'm not sure if exposing it directly vs copy will fuck us in the future towards GPU driven pipelines.

**Comment by rileyzzz** [comment]
Yeah a copy should work.
Also, I just realized we might not have a clean way to submit draw calls using a GPU buffer incase we want to modify the copy, or generate new geometry based off of it.
Seems like the Unity way of doing it is to let you submit a draw call without a vertex buffer and just an explicit index count, and let the vertex shader reconstruct its vertices from the custom buffer.

https://docs.unity3d.com/6000.0/Documentation/ScriptReference/Graphics.RenderPrimitivesIndexed.html

Would be cool to get a Graphics.DrawPrimitives* API to support that, maybe with support for indirect/instanced versions too

**Comment by handsomematt** [comment]
We have indirect drawing already which should work how you want

**Comment by rileyzzz** [comment]
DrawModelInstancedIndirect takes a Model though, which wouldn't make much sense (unless you can create a Model out of GpuBuffers with ModelBuilder?)
I'd need a way to pass the index buffer along to the indirect call.

**Comment by handsomematt** [comment]
Ah yeah we can get some more raw dog ones now that GpuBuffer can represent Vertex and Index buffers too

**Comment by aylaylay** [comment]
Yep I can work on lower level versions
